### PR TITLE
[Conceptual Draft] Params are now fetched from datums rather than lists

### PR DESCRIPTION
**Pull request by PowerfulBacon**
See https://github.com/BeeStation/BeeStation-Hornet/pull/9299/files#diff-90a63f5baa302f52770bc9773f9b737503ac31292eca0657127b020e2de0c8a6 for the important code.

## About The Pull Request

This PR aims to refactor how params are fetched from both Topic and ui_act calls, reworking it so that instead of pulling from a dictionary, you get elements from a datum. This provides methods for getting sanitised values and prevents coders from pulling any unsanitised values from user input which will significantly reduce the chance of making basic mistakes leading to critical href exploits.

As MSO pointed out when I proposed this idea, there are many UIs which take user input and send it straight back into a TGUI, which performs its own sanitisation. This means that if any text is encoded or sanitised before going into TGUI, it will result in it being double encoded causing it to render > characters as &gt;. I thought about this for a long time and came up with a solution that allows for values to be carried around unsanitised in byond, while still providing some safety against the accidental addition of href exploits. See message containers.

## Why these changes are good/necessary?

If a programmer isn't certain of how href explotis work, they can be very easy to accidentally implement. Certain cases of href exploits can be very hard to detect during review, specifically if the sanitisation is not performed where the href is pulled from. In order to achieve a level of safety and make it easy to spot these during review, I believe the following rules should be followed:
- Unsanitised user input should never be passed as a parameter to a function. Sanitisation should occur beforehand
- Unsanitised user input should never be stored on a non-local variable, since any payloads could be accidentally triggered. (Code that uses that variable could be added in the future, meaning a PR can introduce a critical vulnerability without even touching hrefs making these landmines to work with)

Sometimes we have to deal with unsanitised user input in the case of TGUI paper for example, we take an input from the user and pass it straight back unsanitised. We wrap these inside of a contains which cannot be viewed by admins which prevents accidental use. In addition to this, the unsafe message container is a datum which means it cannot be directly used by new functions removing the ability to accidentally introduce a critical vulnerability from storing unsanitised inputs.

This makes it so that it is:
- Easy to review href exploits
- Difficult to make a href exploit without explicitly jumping through the hoops that this PR introduces to make href exploits
- Easy to get sanitised inputs

### Params

Params holds the unsafe parameter list. This datum is protected and the unsafe params variable cannot be viewed by admins, completely preventing an accidental triggering of a payload by modifying any unsafe variables. (These should never be stored anyway, so this shouldn't matter)

**Boolean Handling**

**is_param_equal_to(param, value)**: Returns true if the requested parameter is equal to the provided value, returns false otherwise.

**get_boolean(param)**: Returns the requested parameter as a boolean value, true if the value evaluates to true and false otherwise.

**Text Handling**

**get_sanitised_text(param)**: Returns the requested parameter as sanitised text. This encodes any HTML and removes \n and \r.

**get_encoded_text(param)**: Returns the requested parameter as html encoded text. This will replace things like < with &gt; so that they can be rendered in HTML UIs.

**get_unsanitised_message_container(param)**: Returns the requested parameter as an unsanitised message. This message is inside an unsafe message container which is a datum that allows the message to be carried around inside of byond in a manner that almost eliminates the risk of programmer error. See `message containers`.

**Numeric Handling**

**get_num(param)**: Returns the requested parameter as a numeric value.

**Path Handling**

**get_subtype_path(param, root_type)**: Returns the parameter as a path, so long as that path is a subtype of root_type. This forces you to provide the type that you expect the subtype to be and does not allow a user to get any abritrary type. This is probably the most dangerous method out of all of them, since root_type could be set to /datum to get any time, however that is an easy mistake to spot in review.

**List Handling**

**locate_param(param, list/source_list)**: Uses the requested parameter as a `\ref` to identify a certain datum inside of the source list.

**get_text_in_list(param, list/source_list)**: Returns the param as text, as long as the message exists inside of the source list. This is the way you get selected categories from user input and selections from limited choices. This is probably the most commonly used one.

**Dictionary Handling**

**get_from_lookup(param, list/dictionary)**: Uses the requested parameter as the key for a dictionary, returns the value associated with that key in the provided dictionary.

**get_param_dict(param)**: Returns the requested parameter as a params dictionary. This is useful if your parameter has subparameters.

------

_Function names that provide fetching of parameters are subject to change. More will be added when I find more requirements for these._

### Message Containers

Message containers carry unsafe messages around inside byond, and add a protective bubble around them to prevent misuse.

**get_unsafe_message()**: Gets the message unsanitised. This should only be used for sending it to TGUI, where TGUI will sanitise it itself.

**get_sanitised()**: Returns a sanitised version of the message.

I will expand on these when needed, adding in methods for processing the unsafe message if we need that.

## Testing Photographs and Procedure

Coming soon pending completion

## Changelog
:cl:
refactor: Refactors how parameters are fetched from ui_act calls to reduce the risk of href exploits.
/:cl:

### CODE REVIEW

File: code/controllers/subsystem/pai.dm [hunk 1]
```diff
@@ -93,27 +93,33 @@ SUBSYSTEM_DEF(pai)
...
+	var/name = candidate.get_sanitised_text("name")
+	if(CHAT_FILTER_CHECK(comments))
 		to_chat(usr, "<span class='warning'>Your OOC comment contains prohibited word(s)!</span>")
 		return FALSE
-	if(CHAT_FILTER_CHECK(params["candidate"]["description"]))
+	if(CHAT_FILTER_CHECK(description))
 		to_chat(usr, "<span class='warning'>Your description contains prohibited word(s)!</span>")
 		return FALSE
-	if(CHAT_FILTER_CHECK(params["candidate"]["name"]))
+	if(CHAT_FILTER_CHECK(name)
```
Comment (Crossedfall) [inline code/controllers/subsystem/pai.dm]:
```suggestion
	if(CHAT_FILTER_CHECK(name))
```

File: code/modules/tgui/params.dm [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+/datum/params/proc/get_boolean(param)
+	return !!_unsafe_params[param]
+
+// =============================
+// Text Handling
+// =============================
+
+/// Returns the requested parameter as fully sanitised text, removing \n and \t as well as encoding HTML.
+/datum/params/proc/get_sanitised_text(param)
+	return sanitize_text(_unsafe_params[param])
```
Comment (adamsong) [inline code/modules/tgui/params.dm]:
```suggestion
	return sanitize(_unsafe_params[param])
```

sanitize_text only checks if the input is text

Comment (PowerfulBacon) [reply code/modules/tgui/params.dm]: Ah yea, you are correct. Good spot

**Comment by PowerfulBacon** [comment]
This is going to take an incredibly long time to complete and test all the required changes, please let me know sooner rather than later if there are any major issues that make this an unwanted change.

File: code/modules/library/lib_codex_gigas.dm [hunk 3]
```diff
@@ -64,7 +64,7 @@
 /obj/item/book/codex_gigas/proc/ask_name(mob/reader)
 	ui_interact(reader)
 
-/obj/item/book/codex_gigas/ui_act(action, params)
+/obj/item/book/codex_gigas/ui_act(action, datum/params/params)
```
Comment (adamsong) [inline code/modules/library/lib_codex_gigas.dm:67]: Heads up, `action` contains user input here

**Comment by PowerfulBacon** [comment]
This is probably going to be too much work for me right now, but a good PR. It should work in its current state anyway since I added an override for indexing

**Pull request closed without merging**